File: object-oriented-programming/virtual-world-python/plants/belladonna.py
from plants.plant import Plant
from species import Species
import random

class Belladonna(Plant) :

    def __init__(self, world, x = None, y = None) :
        if x == None and y == None :
            super().__init__(Species.BELLADONNA, 0, 0, 99, world, 1)
            width = world.width
            height = world.height

            while(True):
                new_x = random.randint(0, width - 1)
                new_y = random.randint(0, height - 1)
                if self._world.is_free(new_x, new_y) :
                    break
                else :
                    continue

            self._world.set_organism(None, self.x, self.y)
            self._x = new_x
            self._y = new_y
            self._world.set_organism(self, self.x, self.y)
            return
        
        super().__init__(Species.BELLADONNA, x, y, 99, world, 1)

    def action(self) :
        return super().action()

    def collision(self, other):
         self._world.report(f"Belladonna killed {other.species.display_name()}!")
         self.mark_dead()
         other.mark_dead()
         self._world.set_organism(None, self.x, self.y)
         self._world.set_organism(None, other.x, other.y)

    def draw_color(self):
        return self.species.color
    
    def sow(self, x, y):
        return Belladonna(self._world, x, y)

File: object-oriented-programming/virtual-world-java/src/Game/Model/Plants/Belladonna.java
package Game.Model.Plants;

import Game.Model.Organism;
import Game.Model.Species;
import Game.Model.World;

import java.awt.*;
import java.util.Random;

public class Belladonna extends Plant {
    Random random = new Random();

    public Belladonna(int x, int y, World world){
        super(Species.BELLADONNA, x, y, 99, world, 1);
    }

    public Belladonna(World world){
        super(Species.BELLADONNA, 0, 0, 99, world, 1);
        int width = world.getWidth();
        int height = world.getHeight();

        int newX, newY;

        do {
            newX = random.nextInt(width);
            newY = random.nextInt(height);
        } while(!world.isFree(newY, newX));

        world.setOrganism(null, y, x);
        x = newX;
        y = newY;
        world.setOrganism(this, y, x);
    }

    public void collision(Organism other){
        world.report("Belladonna killed " + other.getSpecies().displayName() + "!");
        markDead();
        other.markDead();
        world.setOrganism(null, y, x);
        world.setOrganism(null, other.getY(), other.getX());
    }

    public final Color draw(){
        return species.displayColor();
    }

    public Belladonna sow(int y, int x){
        return new Belladonna(x, y, world);
    }
}
